toplevel/role/role.go

```diff
@@ -221,6 +228,61 @@ func (c config) Apply(address string, entriesBytes []byte, dryRun bool, threadPo
 	return nil
 }
 
+// Extracts names of policies referenced within applicable properties of desired roles
+// and updates those properties to only include the names of policies.
+// This is necessary to support policy file references within schemas.
+// The graphql query will return nested object for policy properties but vault api
+// expects only names of policies to be included
+func formatPolicyRefs(desiredRoles []entry) error {
+	// defines applicable properties(s) for each role type to reformat
+	// see https://github.com/app-sre/qontract-schemas/blob/main/schemas/vault-config/role-1.yml
+	typeToProperties := map[string][]string{
+		"approle": {
+			"token_policies",
+			"policies",
+		},
+		"oidc": {
+			"token_policies",
+		},
+		"kubernetes": {
+			"token_policies",
+		},
+	}
+	extracted := []string{}
+	for _, role := range desiredRoles {
+		for _, property := range typeToProperties[role.Type] {
+			policies, ok := role.Options[property].([]interface{})
+			if !ok {
+				return fmt.Errorf("[Vault Role] Failed to convert policy list `%s` within `%s`",
+					property,
+					role.Name,
+				)
+			}
+			for _, policy := range policies {
+				policyMap, ok := policy.(map[interface{}]interface{})
+				if !ok {
+					return fmt.Errorf("[Vault Role] Failed to convert policy map within `%s` attribute of `%s`",
+						property,
+						role.Name,
+					)
+				}
+				policyName, ok := policyMap["name"].(string)
+				if !ok {
+					return fmt.Errorf("[Vault Role] Failed to retrieve policy name within `%s` attribute of `%s`",
+						property,
+						role.Name,
+					)
+				}
+				extracted = append(extracted, policyName)
+			}
+			// overwrite list of policy objs with list of policy names
+			role.Options[property] = extracted
+			extracted = nil
```

Comment:
This variable can have a reduced scope. You only really need it within the second loop body if my weary eyes can see correctly. :smile: 

You can create a new slice there every time you need it, and there wouldn't be a need to clear it. The GC will take care of the old allocations for you in this case once these either aren't needed or things go out of scope.

Reply:
Potentially silly question related to this one. I consciously placed the `extracted` var definition in its current position because I thought the broader scope was a worthwhile trade for any performance impact by creating the slice on each iteration with the loop. Is repetitive variable creation a

Reply:
@dwelch0, great question! Let's have a look how this looks like.

Consider this little benchmark suite:

```go
package main

import (
	"testing"
)

func A() {
	m := map[int][]string{}

	for i, _ := range []int{1, 2, 3} {
		var x []string
		for _, c := range []string{"a", "b", "c"} {
			x = append(x, c)
		}
		m[i] = x
	}
}

func B() {
	m := map[int][]string{}

	var x []string
	for i, _ := range []int{1, 2, 3} {
		for _, c := range []string{"a", "b", "c"} {
			x = append(x, c)
		}
		m[i] = x
		x = nil
	}
}

func BenchmarkA(b *testing.B) {
	for i := 0; i < b.N; i++ {
		A()
	}
}

func BenchmarkB(b *testing.B) {
	for i := 0; i < b.N; i++ {
		B()
	}
}
```

Now, when run, yields:

```bash
$ go test -bench=. -benchmem 
goos: linux
goarch: amd64
pkg: test
cpu: AMD Ryzen 9 6900HX with Radeon Graphics        
BenchmarkA-16    	 1637730	       763.1 ns/op	     336 B/op	       9 allocs/op
BenchmarkB-16    	 1480376	       789.4 ns/op	     336 B/op	       9 allocs/op
PASS
ok  	test	3.981s
```

The results are fascinating.

Different, a more "real life" benchmark, yields:

![Screenshot from 2023-08-18 04-28-31](https://github.com/app-sre/vault-manager/assets/250723/2b21a622-6410-433d-855b-ff1e0329bac3)

Also interesting! But why?

We can compare the actual code behind both versions:

![Screenshot from 2023-08-18 04-59-49](https://github.com/app-sre/vault-manager/assets/250723/6adbe04d-fd60-493c-afb0-cba1c2842629)

The binaries are not the same:

```bash
$ sha1sum main_A main_B
8be856ba2b871adc0995d726642028ef3ed4de23  main_A
063fde6a8a2b9cfb8169d78100ab829b8aaab982  main_B
```

The assembly looks identical! So where is this tiny difference in performance coming from?

Perhaps when and how GC runs? Let's check if our code escapes to the heap per:

```bash
$ go run -gcflags '-m -l' main_A.go 
# command-line-arguments
./main_A.go:4:23: map[int][]string{} does not escape
./main_A.go:6:25: []int{...} does not escape
./main_A.go:8:29: []string{...} does not escape

$ go run -gcflags '-m -l' main_B.go 
# command-line-arguments
./main_B.go:4:23: map[int][]string{} does not escape
./main_B.go:7:25: []int{...} does not escape
./main_B.go:8:29: []string{...} does not escape
```

We are all allocated on the stack - as expected. So, how about the GC? This requires a small change to our test code so that we can collect meaningful sample data:

```go
package main

import (
	"runtime"
	"time"
)

func A() {
	m := map[int][]string{}

	for i, _ := range []int{1, 2, 3} {
		var x []string
		for _, c := range []string{"a", "b", "c"} {
			x = append(x, c)
		}
		m[i] = x
	}
	runtime.GC()

}

func main() {
	A()
	runtime.GC()
	time.Sleep(time.Second * 2)
}
```

(the "B" has been modified accordingly)

This is what the GC looks like for both:

```bash
$ time GODEBUG=gctrace=1 go run main_B.go 
gc 1 @0.006s 1%: 0.017+0.26+0.018 ms clock, 0.28+0.059/0.65/0.045+0.28 ms cpu, 3->3->0 MB, 4 MB goal, 0 MB stacks, 0 MB globals, 16 P
gc 2 @0.011s 7%: 0.83+1.7+0.025 ms clock, 13+0/2.5/0+0.40 ms cpu, 4->4->1 MB, 4 MB goal, 0 MB stacks, 0 MB globals, 16 P
gc 3 @0.014s 8%: 0.078+0.46+0.035 ms clock, 1.2+0.42/0.83/0.022+0.57 ms cpu, 3->4->1 MB, 4 MB goal, 0 MB stacks, 0 MB globals, 16 P
gc 4 @0.016s 8%: 0.018+0.12+0.001 ms clock, 0.30+0.077/0.42/0.50+0.030 ms cpu, 3->3->0 MB, 4 MB goal, 0 MB stacks, 0 MB globals, 16 P
gc 5 @0.017s 7%: 0.005+0.19+0.002 ms clock, 0.093+0.068/0.60/0.48+0.042 ms cpu, 3->3->0 MB, 4 MB goal, 0 MB stacks, 0 MB globals, 16 P
gc 6 @0.018s 7%: 0.005+0.22+0.010 ms clock, 0.092+0.048/0.51/0.40+0.16 ms cpu, 3->3->1 MB, 4 MB goal, 0 MB stacks, 0 MB globals, 16 P
gc 7 @0.019s 7%: 0.009+0.17+0.003 ms clock, 0.15+0.081/0.48/0.42+0.056 ms cpu, 3->3->0 MB, 4 MB goal, 0 MB stacks, 0 MB globals, 16 P
# command-line-arguments
gc 1 @0.000s 6%: 0.004+0.43+0.002 ms clock, 0.071+0.013/0.91/0+0.037 ms cpu, 6->7->6 MB, 6 MB goal, 0 MB stacks, 0 MB globals, 16 P
gc 2 @0.007s 1%: 0.007+0.57+0.009 ms clock, 0.11+0/1.1/0.47+0.14 ms cpu, 12->12->8 MB, 13 MB goal, 0 MB stacks, 0 MB globals, 16 P
gc 1 @0.000s 8%: 0.033+0.21+0.004 ms clock, 0.53+0/0.095/0+0.069 ms cpu, 0->0->0 MB, 4 MB goal, 0 MB stacks, 0 MB globals, 16 P (forced)
gc 2 @0.000s 12%: 0.047+0.060+0.003 ms clock, 0.76+0/0.069/0.003+0.059 ms cpu, 0->0->0 MB, 4 MB goal, 0 MB stacks, 0 MB globals, 16 P (forced)

real	0m2.052s
user	0m0.062s
sys	0m0.081s

$ time GODEBUG=gctrace=1 go run main_A.go 
gc 1 @0.008s 0%: 0.016+0.47+0.012 ms clock, 0.26+0.082/0.47/0.12+0.19 ms cpu, 3->3->0 MB, 4 MB goal, 0 MB stacks, 0 MB globals, 16 P
gc 2 @0.011s 4%: 0.31+3.4+0.015 ms clock, 5.0+0.010/3.4/0+0.24 ms cpu, 3->4->1 MB, 4 MB goal, 0 MB stacks, 0 MB globals, 16 P
gc 3 @0.015s 4%: 0.050+0.25+0.002 ms clock, 0.81+0.28/0.61/0.007+0.046 ms cpu, 3->4->1 MB, 4 MB goal, 0 MB stacks, 0 MB globals, 16 P
gc 4 @0.016s 4%: 0.010+0.15+0.002 ms clock, 0.17+0.072/0.44/0.49+0.034 ms cpu, 3->3->1 MB, 4 MB goal, 0 MB stacks, 0 MB globals, 16 P
gc 5 @0.017s 4%: 0.006+0.14+0.007 ms clock, 0.099+0.025/0.39/0.34+0.12 ms cpu, 3->3->0 MB, 4 MB goal, 0 MB stacks, 0 MB globals, 16 P
gc 6 @0.019s 4%: 0.004+0.13+0.012 ms clock, 0.077+0.026/0.36/0.37+0.20 ms cpu, 3->3->1 MB, 4 MB goal, 0 MB stacks, 0 MB globals, 16 P
gc 7 @0.019s 4%: 0.005+0.15+0.009 ms clock, 0.085+0.027/0.42/0.59+0.15 ms cpu, 3->3->1 MB, 4 MB goal, 0 MB stacks, 0 MB globals, 16 P
# command-line-arguments
gc 1 @0.000s 6%: 0.006+0.55+0.011 ms clock, 0.10+0.012/1.0/0+0.19 ms cpu, 6->7->6 MB, 6 MB goal, 0 MB stacks, 0 MB globals, 16 P
gc 2 @0.007s 2%: 0.007+0.73+0.020 ms clock, 0.11+0/1.2/0.42+0.33 ms cpu, 12->12->8 MB, 13 MB goal, 0 MB stacks, 0 MB globals, 16 P
gc 1 @0.000s 3%: 0.006+0.068+0.003 ms clock, 0.10+0/0.10/0.005+0.062 ms cpu, 0->0->0 MB, 4 MB goal, 0 MB stacks, 0 MB globals, 16 P (forced)
gc 2 @0.000s 8%: 0.035+0.055+0.003 ms clock, 0.56+0/0.10/0.014+0.056 ms cpu, 0->0->0 MB, 4 MB goal, 0 MB stacks, 0 MB globals, 16 P (forced)

real	0m2.053s
user	0m0.072s
sys	0m0.067s
```

Hard to tell, to be honest, as the small sample (of one) does not help here. And even when you run the benchmark code for each function and measure GC outcomes per:

```bash
$ GODEBUG=gctrace=1 go test -bench=. -benchmem -run 'BenchmarkA' >/dev/null
gc 1 @0.006s 1%: 0.022+0.22+0.016 ms clock, 0.36+0.13/0.55/0.099+0.25 ms cpu, 3->4->0 MB, 4 MB goal, 0 MB stacks, 0 MB globals, 16 P
gc 2 @0.012s 2%: 0.16+0.69+0.016 ms clock, 2.6+0.78/1.2/0+0.26 ms cpu, 3->4->1 MB, 4 MB goal, 0 MB stacks, 0 MB globals, 16 P
gc 3 @0.013s 3%: 0.054+0.31+0.007 ms clock, 0.86+0.11/0.82/0.16+0.11 ms cpu, 3->3->1 MB, 4 MB goal, 0 MB stacks, 0 MB globals, 16 P
gc 4 @0.015s 4%: 0.044+0.53+0.044 ms clock, 0.71+0.49/0.97/0.001+0.71 ms cpu, 3->4->1 MB, 4 MB goal, 0 MB stacks, 0 MB globals, 16 P
gc 5 @0.016s 5%: 0.069+0.43+0.017 ms clock, 1.1+0.49/1.1/0.087+0.27 ms cpu, 3->3->1 MB, 4 MB goal, 0 MB stacks, 0 MB globals, 16 P
gc 6 @0.018s 4%: 0.041+3.6+0.016 ms clock, 0.66+0.50/2.2/0+0.26 ms cpu, 3->4->2 MB, 4 MB goal, 0 MB stacks, 0 MB globals, 16 P
gc 7 @0.023s 5%: 0.076+0.84+0.051 ms clock, 1.2+0.68/2.2/0.006+0.82 ms cpu, 4->5->3 MB, 5 MB goal, 0 MB stacks, 0 MB globals, 16 P
gc 8 @0.025s 6%: 0.032+0.38+0.011 ms clock, 0.51+0.53/1.1/1.1+0.18 ms cpu, 5->6->2 MB, 6 MB goal, 0 MB stacks, 0 MB globals, 16 P
gc 9 @0.026s 6%: 0.007+0.28+0.022 ms clock, 0.11+0.043/0.91/0.71+0.36 ms cpu, 4->5->2 MB, 4 MB goal, 0 MB stacks, 0 MB globals, 16 P
gc 10 @0.026s 6%: 0.006+0.19+0.004 ms clock, 0.10+0.050/0.64/0.81+0.074 ms cpu, 4->4->2 MB, 5 MB goal, 0 MB stacks, 0 MB globals, 16 P
gc 11 @0.027s 6%: 0.006+0.21+0.010 ms clock, 0.10+0.055/0.63/1.1+0.16 ms cpu, 3->4->2 MB, 4 MB goal, 0 MB stacks, 0 MB globals, 16 P
gc 12 @0.028s 6%: 0.005+0.19+0.001 ms clock, 0.094+0.028/0.68/1.4+0.029 ms cpu, 3->4->2 MB, 4 MB goal, 0 MB stacks, 0 MB globals, 16 P
gc 13 @0.029s 6%: 0.018+0.37+0.003 ms clock, 0.29+0.42/1.1/0.20+0.062 ms cpu, 3->5->2 MB, 4 MB goal, 0 MB stacks, 0 MB globals, 16 P
gc 14 @0.030s 6%: 0.021+0.35+0.032 ms clock, 0.33+0.41/0.99/0.85+0.52 ms cpu, 4->5->2 MB, 5 MB goal, 0 MB stacks, 0 MB globals, 16 P
gc 15 @0.031s 6%: 0.016+0.23+0.014 ms clock, 0.26+0.35/0.70/0.39+0.23 ms cpu, 4->5->2 MB, 5 MB goal, 0 MB stacks, 0 MB globals, 16 P
gc 16 @0.031s 6%: 0.012+0.21+0.007 ms clock, 0.20+0.075/0.66/1.2+0.11 ms cpu, 5->5->2 MB, 6 MB goal, 0 MB stacks, 0 MB globals, 16 P
gc 17 @0.032s 7%: 0.022+0.32+0.059 ms clock, 0.35+0.11/0.73/0.97+0.95 ms cpu, 4->5->2 MB, 4 MB goal, 0 MB stacks, 0 MB globals, 16 P
gc 18 @0.033s 7%: 0.054+0.95+0.010 ms clock, 0.87+0.34/2.2/0+0.16 ms cpu, 4->5->3 MB, 5 MB goal, 0 MB stacks, 0 MB globals, 16 P
gc 19 @0.035s 8%: 0.24+0.57+0.011 ms clock, 3.8+0.40/1.7/0.44+0.18 ms cpu, 5->6->3 MB, 6 MB goal, 0 MB stacks, 0 MB globals, 16 P
# test.test
gc 1 @0.000s 7%: 0.019+0.32+0.002 ms clock, 0.31+0.014/0.67/0+0.045 ms cpu, 6->7->6 MB, 6 MB goal, 0 MB stacks, 0 MB globals, 16 P
gc 2 @0.002s 4%: 0.005+0.44+0.005 ms clock, 0.082+0.046/1.0/0+0.089 ms cpu, 13->13->12 MB, 13 MB goal, 0 MB stacks, 0 MB globals, 16 P
gc 3 @0.033s 1%: 0.055+0.77+0.016 ms clock, 0.88+0.022/2.2/0+0.27 ms cpu, 23->24->12 MB, 24 MB goal, 0 MB stacks, 0 MB globals, 16 P

$ GODEBUG=gctrace=1 go test -bench=. -benchmem -run 'BenchmarkB' >/dev/null
gc 1 @0.006s 1%: 0.018+0.36+0.033 ms clock, 0.29+0.13/0.87/0.025+0.54 ms cpu, 3->4->0 MB, 4 MB goal, 0 MB stacks, 0 MB globals, 16 P
gc 2 @0.011s 3%: 0.16+0.66+0.017 ms clock, 2.5+0.62/1.2/0.077+0.27 ms cpu, 3->4->1 MB, 4 MB goal, 0 MB stacks, 0 MB globals, 16 P
gc 3 @0.014s 3%: 0.026+0.22+0.010 ms clock, 0.42+0.11/0.69/0.79+0.16 ms cpu, 3->3->1 MB, 4 MB goal, 0 MB stacks, 0 MB globals, 16 P
gc 4 @0.015s 4%: 0.047+0.63+0.065 ms clock, 0.75+0.62/1.2/0.085+1.0 ms cpu, 3->4->1 MB, 4 MB goal, 0 MB stacks, 0 MB globals, 16 P
gc 5 @0.017s 4%: 0.022+3.4+0.010 ms clock, 0.36+1.0/1.3/0+0.17 ms cpu, 3->5->2 MB, 4 MB goal, 0 MB stacks, 0 MB globals, 16 P
gc 6 @0.022s 5%: 0.095+0.69+0.006 ms clock, 1.5+1.2/1.9/0+0.098 ms cpu, 4->5->2 MB, 4 MB goal, 0 MB stacks, 0 MB globals, 16 P
gc 7 @0.023s 5%: 0.032+0.44+0.011 ms clock, 0.52+0.18/1.3/0.79+0.17 ms cpu, 5->6->2 MB, 6 MB goal, 0 MB stacks, 0 MB globals, 16 P
gc 8 @0.024s 5%: 0.012+0.23+0.001 ms clock, 0.19+0.15/0.83/0.60+0.028 ms cpu, 5->5->2 MB, 6 MB goal, 0 MB stacks, 0 MB globals, 16 P
gc 9 @0.025s 5%: 0.006+0.24+0.009 ms clock, 0.10+0.058/0.68/1.1+0.15 ms cpu, 4->4->2 MB, 5 MB goal, 0 MB stacks, 0 MB globals, 16 P
gc 10 @0.026s 5%: 0.005+0.26+0.024 ms clock, 0.084+0.025/0.81/1.1+0.38 ms cpu, 3->4->2 MB, 4 MB goal, 0 MB stacks, 0 MB globals, 16 P
gc 11 @0.027s 5%: 0.006+0.23+0.007 ms clock, 0.10+0.032/0.59/1.1+0.12 ms cpu, 3->4->2 MB, 4 MB goal, 0 MB stacks, 0 MB globals, 16 P
gc 12 @0.028s 6%: 0.019+0.31+0.037 ms clock, 0.31+0.46/0.79/0.065+0.60 ms cpu, 3->5->2 MB, 4 MB goal, 0 MB stacks, 0 MB globals, 16 P
gc 13 @0.028s 6%: 0.047+0.31+0.021 ms clock, 0.76+0.18/0.95/0.071+0.34 ms cpu, 4->6->2 MB, 5 MB goal, 0 MB stacks, 0 MB globals, 16 P
gc 14 @0.029s 6%: 0.017+0.23+0.004 ms clock, 0.28+0.092/0.79/0.70+0.071 ms cpu, 5->6->2 MB, 6 MB goal, 0 MB stacks, 0 MB globals, 16 P
gc 15 @0.030s 6%: 0.005+0.20+0.002 ms clock, 0.089+0.032/0.60/1.1+0.046 ms cpu, 5->5->2 MB, 5 MB goal, 0 MB stacks, 0 MB globals, 16 P
gc 16 @0.031s 6%: 0.019+0.53+0.005 ms clock, 0.30+0.42/1.1/0.037+0.092 ms cpu, 4->6->3 MB, 5 MB goal, 0 MB stacks, 0 MB globals, 16 P
gc 17 @0.032s 7%: 0.14+1.6+0.065 ms clock, 2.3+0.14/2.6/0+1.0 ms cpu, 5->7->3 MB, 6 MB goal, 0 MB stacks, 0 MB globals, 16 P
# test.test
gc 1 @0.000s 5%: 0.021+0.49+0.006 ms clock, 0.34+0.013/0.55/0.075+0.10 ms cpu, 6->7->6 MB, 6 MB goal, 0 MB stacks, 0 MB globals, 16 P
gc 2 @0.002s 4%: 0.004+0.63+0.005 ms clock, 0.076+0.047/0.96/0+0.080 ms cpu, 13->13->12 MB, 13 MB goal, 0 MB stacks, 0 MB globals, 16 P
gc 3 @0.030s 0%: 0.014+0.49+0.009 ms clock, 0.23+0/1.2/0.71+0.15 ms cpu, 23->23->12 MB, 24 MB goal, 0 MB stacks, 0 MB globals, 16 P
```

There does not seem to be that much of a difference.

Given that the assembly code is the same, I am not sure why "A" runs slightly faster _most of the time_.

A few things to keep in mind:

- The test code here is tiny and does not reflect what the Vault Manager is really doing
- These synthetic benchmarks can yield inconclusive outcomes as the machine tests run on has different ambience for each test
- For a small _n_, unless you are leaking memory or things that are grossly inefficient, the answer often is "it does not matter", as the compiler and the runtime would optimise things behind the scenes anwyay
- Benchmarking is hard! :smile:

Reply:
@dwelch0, there would be no difference in performance.

The resulting code (assembly) is the same and simply a matter of common patterns and code readability.